### ISSUE TITLE
Can't use subexpression in partial parameter

### ISSUE THREAD

**Issue by mohd-akram** [rexm/Handlebars.Net#208]
So I'm trying to do something like this:

`{{> head title=(_ "Sign Up") }}`

where `_` is a helper I've registered, but I'm getting this error:

`HandlebarsDotNet.HandlebarsCompilerException: Token ')' could not be converted to an expression`

This works in the JavaScript library:

```javascript
Handlebars.registerHelper('_', function(s) {
	return new Handlebars.SafeString(s);
});
Handlebars.registerPartial('head', 'title is "{{title}}"');
var template = Handlebars.compile('{{> head title=(_ "Sign Up") }}');
alert(template());
```

**Comment by rexm** [comment]
The current implementation of the hash parameters parser is very simple... it doesn't support anything but literals in the value position. We need to remove the parsing logic from the WordParser and move it to the core lexer, then the other tokenizers like the SubExpression parser will be able to pick up the `()` and interpret the contents.

I may get to this in the next few weeks. PR always welcome in the meantime!

**Comment by pmccloghrylaing** [comment]
I'm interested in helping with this.

The tokenizing should be simple to update - instead of treating `param=value` as a token, it can be broken into three tokens with `=` as it's own token.
The converters are more difficult, I wonder if parameters and subexpressions need to be handled in the same converter? It's difficult because hash parameters can exist in subexpressions and vice versa (eg. `{{>partial item1=(helper item1="val1") item2="val2"}}`).
@rexm how were you thinking about handling this?

**Comment by rexm** [comment]
The extent of my thinking so far is in the comment above - anything deeper is welcome. Could you explain the one you started?

**Comment by pmccloghrylaing** [comment]
**Converters/Accumulators**
I had found I needed to convert to hash parameters before and after the `SubExpressionConverter`. The `HashParametersConverter` would skip hash parameters if the next token was a `StartSubExpressionToken`. I've been thinking more about this and I don't think it's ideal as it won't support deeply nested combinations of sub-expressions and parameters. I'm currently thinking it would be worth handling parameters with an interim expression for the parameter assignment excluding the value (`param=`), that the `SubExpressionConverter` isn't going to throw on, and then run an accumulator after the `SubExpressionConverter` to pick up the value.

**Binders**
With the binders I've looked some more into it and returning a `BlockExpression` for creating the parameter dictionary seems like a good way to go as it'll return the result of the last expression, so something like this will do what we want it to do:
```
var x = new HashParameterDictionary();
x.Add(key, value);
...
x;
```

**Visitors**
The other main change I was looking at was a refactor of the Visitors, starting with the base `HandlebarsExpressionVisitor`. The [System.Linq.Expressions.ExpressionVisitor](https://github.com/Microsoft/referencesource/blob/master/System.Core/System/Linq/Expressions/ExpressionVisitor.cs) methods follow a pattern of calling `Visit()` on all child Expressions and creating a new expression if they differ, ensuring any changes will be propagated. Applying this pattern to the `HandlebarsExpressionVisitor` means fewer overrides need to be written in the sub-classes.

**Comment by pmccloghrylaing** [comment]
I took another look at the Binder and creating the equivalent of `new HashParameterDictionary { { "key", value }, ... }` is trivial.

**Comment by rexm** [comment]
So to make sure I am following - we need to introduce a more sophisticated notion of a hash parameter where the right side of each parameter (the value) is an expression that could either be a const or a subexpression, and hash parameter converter should assume the right side has already been converted/processed, but it is not sufficient to simply run the subexpression converter/accumulator before hash parameters converter runs?

**Comment by pmccloghrylaing** [comment]
I've opened a PR (#262) with my approach to it - I've got it working using the two converters:
- The first one converts a `WordExpressionToken + AssignmentToken` into a `HashParameterAssignmentExpression` before the SubExpressionConverter (and HelperConverter) runs.
- The second one accumulates `(HashParameterAssignmentExpression + Expression)+` into a `HashParametersExpression`. It needs to Visit SubExpressionExpressions & HelperExpressions in order to update handle nested parameters.

The converter changes can be seen here: https://github.com/rexm/Handlebars.Net/pull/262/commits/3f972b057725dd8ddd8a77f018b541ffc3ba288e
